### PR TITLE
PUT Blob Structured Message

### PR DESCRIPTION
**Pull request by jaschrep-msft**
Adds structured message to PUT blob on block blob uploads.

### CODE REVIEW

File: sdk/storage/Azure.Storage.Blobs/src/BlockBlobClient.cs [hunk 1]
```diff
@@ -875,14 +875,36 @@ public new BlockBlobClient WithEncryptionScope(string encryptionScope)
...
-                    content = content?.WithNoDispose().WithProgress(progressHandler);
+                    ContentHasher.GetHashResult hashResult = null;
+                    long contentLength = (content?.Length - content?.Position) ?? 0;
+                    long? structuredContentLength = default;
+                    string structuredBodyType = null;
+                    if (validationOptions != null &&
+                        validationOptions.ChecksumAlgorithm.ResolveAuto() == StorageChecksumAlgorithm.StorageCrc64 &&
+                        validationOptions.PrecalculatedChecksum.IsEmpty &&
+                        ClientSideEncryption == null) // don't allow feature combination
+                    {
```
Comment (jalauzon-msft) [inline sdk/storage/Azure.Storage.Blobs/src/BlockBlobClient.cs:886]: Do we still want to use SM if `content` is `null` or `contentLength` is 0?

File: sdk/storage/Azure.Storage.Blobs/src/BlockBlobClient.cs [hunk 2]
```diff
@@ -875,14 +875,36 @@ public new BlockBlobClient WithEncryptionScope(string encryptionScope)
...
+                        // report progress in terms of caller bytes, not encoded bytes
+                        structuredContentLength = contentLength;
+                        contentLength = (content?.Length - content?.Position) ?? 0;
```
Comment (jalauzon-msft) [inline sdk/storage/Azure.Storage.Blobs/src/BlockBlobClient.cs:889]: I believe this is the same as it is when canulated on line 879 so this line is redundant?

File: sdk/storage/Azure.Storage.Blobs/src/BlockBlobClient.cs [hunk 3]
```diff
@@ -875,14 +875,36 @@ public new BlockBlobClient WithEncryptionScope(string encryptionScope)
...
+                        structuredBodyType = Constants.StructuredMessage.CrcStructuredMessage;
+                        content = content.WithNoDispose().WithProgress(progressHandler);
+                        content = new StructuredMessageEncodingStream(
+                            content,
+                            Constants.StructuredMessage.DefaultSegmentContentLength,
+                            StructuredMessage.Flags.StorageCrc64);
+                        contentLength = (content?.Length - content?.Position) ?? 0;
```
Comment (jalauzon-msft) [inline sdk/storage/Azure.Storage.Blobs/src/BlockBlobClient.cs:896]: nit: Do we need null checks here?

**Review by jalauzon-msft — APPROVED** [review]

File: sdk/storage/Azure.Storage.Blobs/src/BlockBlobClient.cs (hunk 1 above)
Comment (jaschrep-msft) [reply sdk/storage/Azure.Storage.Blobs/src/BlockBlobClient.cs:886]: probably not. probably also want to go back and prevent this on stage block.

File: sdk/storage/Azure.Storage.Blobs/src/BlockBlobClient.cs (hunk 3 above)
Comment (jaschrep-msft) [reply sdk/storage/Azure.Storage.Blobs/src/BlockBlobClient.cs:896]: not after we filter the original null out of this if block, but it seemed necessary based on previous code.

**Pull request merged**